lib/galaxy/tools/parameters/basic.py

```diff
@@ -845,13 +843,12 @@ def get_legal_values(self, trans, other_values):
             return self.legal_values
 
     def from_json(self, value, trans, other_values={}):
+        if not value and not self.optional:
+            raise ValueError("An invalid option was selected for %s, please verify." % (self.name))
+        if not value:
+            return None
         legal_values = self.get_legal_values(trans, other_values)
-        workflow_building_mode = trans.workflow_building_mode
-        for context_value in other_values.values():
-            if is_runtime_value(context_value):
-                workflow_building_mode = workflow_building_modes.ENABLED
-                break
-        if len(list(legal_values)) == 0 and workflow_building_mode:
+        if len(list(legal_values)) == 0:
```

Comment:
I don't think the cast to `list` is necessary, `legal_values` should always be a `list` or a `set`.

Reply:
`legal_values` are sets converted to lists, I believe this avoids option duplication.

Reply:
`set()` of a list would remove duplicated values, not the other way round. Any way duplications don't matter here, since we are just checking if it is empty or not.

Reply:
The `sets` can not contain duplicates. The conversion to list only happens to determine the number of non-duplicate entries. I'll check if checking for `None` is enough here.

Reply:
`if len(legal_values) == 0:` or `if legal_values:` should both do what you want.

Reply:
```
In [1]: legal_values = set([1, 2, 3, 2])

In [2]: legal_values
Out[2]: {1, 2, 3}

In [3]: len(legal_values) == 0
Out[3]: False

In [4]: len(list(legal_values))
Out[4]: 3

In [5]: len(list(legal_values)) == 0
Out[5]: False

In [6]: not legal_values
Out[6]: False

In [7]: legal_values = set()

In [8]: len(legal_values) == 0
Out[8]: True

In [9]: len(list(legal_values)) == 0
Out[9]: True

In [10]: not legal_values
Out[10]: True
```